### ISSUE TITLE
Trial's decorate makes the Twisted test suite much slower when using --force-gc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3034]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3034|
|Type|defect|
|Created|2008-02-14 14:54:58Z|
|Branch|https://github.com/twisted/twisted/tree/trial-performance-3034-3|

r22125 increased the runtime of `trial --force-gc twisted` by about a factor of 4.

_Data goes here_.

r22125 also noticably slowed down the Twisted test suite _without_ `--force-gc`. It now takes about 1.2 times as long. Experiments indicate that this is due to Trial keeping more references to tests than it used to.

Attachments:

* [test-plan.jpg](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/622/62263de506e0996f5d6fcb2d814f68c672192bf4/261dd04a9eba82c0af4c6872280885c4096ad602.jpg) (316176 bytes) - added by jml on 2008-03-22 01:31:14Z - I'm a bastard
* [full2.5-debian.stats](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/622/62263de506e0996f5d6fcb2d814f68c672192bf4/5e0e57e61e77a106a7d730cd820b6998690f2190.stats) (11016 bytes) - added by exarkun on 2008-03-23 01:42:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3034 3034
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_trial_performance_3034_3 branches/trial-performance-3034-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1203000898000000 1203000898000000
changetime__1208878674000000 1208878674000000
version__None None
owner__ 
cc__spiv cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

For those playing along at home, r22125 changes Trial to decorate all tests by adapting them to `ITestCase`. For the vast majority of tests, this should be a noop.

I don't care about `--force-gc` speed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22551]) Branching to 'trial-performance-3034'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I've found a solution in r22552. Basically, we don't destroy the hold suite in decorate, so it keeps a reference to all tests. The reference to the suite is kept in script.trial.

jml, do you think it's an acceptable solution? It could be cleaner if `TestSuite` had a method to remove registered tests. Another solution might be to inline the _getSuite call in script.trial to not keep a reference to the suite.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [22706]) Add a test and modify docstring.

Refs [#3034](https://github.com/twisted/twisted/issues/3034)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spiv set owner to @therve|
|-|-|

So the theory here is simply that holding a few thousand less objects in memory makes the test suite _twice_ as fast?  That seems pretty bizarre to me.

(I can understand it for the `--force-gc` case, because it makes sense that more live objects would make the garbage collector slower, but not the normal case.  Surely garbage collection is not taking the majority of the time in a typical test run?)

therve, do you have some timings with and without your proposed fix?

As far as the code itself:

 * "suppose" should be "supposes".  Even better would be "assumes", I think.
 * I don't quite follow the docstring on the test.  What does "it" refer to?    I think the phrasing you are after is "When decorating a test suite, the number of references to the test objects in that test suite should stay the same."
 * The test is testing a particular behaviour (that refcounts of certain objects stay the same), but there's no explanation in either the test or the code of _why_ this is an interesting or desirable property for the code to have.  At both places (the code implementing this behaviour and this test), it should be explicit why we want to be so careful about references.  But as I've already said, it seems strange to me that it should matter so much.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|spiv commented|
|-|-|

Ok, so from IRC, here's the theory:

We try to free test objects after running them.  So that's why holding an extra reference matters: it allows us to free the test case object after they are run.  It shouldn't make any difference at the point the code in this patch is run, but it should make a difference after that point, as the code is run.

Relying on freeing memory like this is a little bit prone to problems, though.  It's _very_ easy to keep references live by accident.  So, while I think this patch is ok, I think there should also be a second layer of defence against this problem: after running a successful test and before dropping our (presumed) last reference to it, clear its `__dict__`.

The reason I suggest that is that I don't think the main problem is a few thousand TestCase objects.  Rather, I think it's the many more thousands of objects that _they_ reference.  So if we clear their instance variables, even if we keep the `TestCase`s object itself live, we'll get 99% of the benefit.  e.g. I'd expect a GUI runner might keep its own reference to tests, so doing this would bring the speed/memory benefits to that GUI runner, where your patch wouldn't.

And at the very least, the analysis of the problem needs to be in a comment next to the code that purports to fix it, so that people have some idea why the code is the way it is.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [22711]) Add some docstring and comments to clarify our purpose.

Refs [#3034](https://github.com/twisted/twisted/issues/3034)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

I added some comments to try to explain the code, and applied the corrections you gave me.

I agree that cleaning up test ```__dict__``` is a good thing to do. I'm not sure this should be done here though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

The `__dict__` clearing thing is a somewhat hackish workaround.  Mangling the objects while they remain in memory is a little bit rude, although it _probably_ wouldn't cause any _practical_ problems.  It's only necessary if we don't successfully forget about `TestCase` instances, so we should just forget successfully.  I feel like there could be a better test for this, maybe something using higher-level trial APIs.

I think the `_tests` mangling needs to go through a public API.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @jml|
|-|-|

I didn't a find a way to add the API. If someone else wants to give a try, he's welcome.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

It'd be really helpful if someone could show me how to reproduce this. In particular:

 * Which tests to run
 * What code to change that induces a significant change in test run time.

On my laptop, the whole suite takes about 2.5 minutes -- with or without decoration.

Not only would this help in solving the problem, it might also provide criteria for closing the ticket.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

This is mainly visible with running the full test suite with forced garbage collection. You can see some numbers on the OS X slave.
 * on trunk: http://buildbot.twistedmatrix.com/builders/osx-py2.4-select-gc/builds/556, 4704s
 * with the fix: http://buildbot.twistedmatrix.com/builders/osx-py2.4-select-gc/builds/555, 1375s

Some numbers on my machine:
 * twisted.test: 46s on trunk, 39s with the fix
 * twisted.test with force-gc: 180s on trunk, 160s with the fix

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml commented|
|-|-|


 * It seems that "factor of 2" is an overstatement.
 * "The fix" is the fix in the branch linked from the ticket?
 * I don't think we should worry about the --force-gc runtme.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Hi,

Notice that therve's numbers are for `twisted.test`, whereas the performance numbers mentioned in the ticket description are for `twisted`.  It shouldn't be at all surprising that the behavior is non-linear.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml commented|
|-|-|

Then I'm still stuck, unable to reproduce this problem.

As I mentioned earlier, running the full test suite produces no appreciable difference on my laptop.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

The last time I looked at the branch, it sped up the test suite by about 2x.  It doesn't now, though.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_18"></a>|@jml commented|
|-|-|

Some numbers:

| Twisted Version | Run 1 | Run 2 | Run 3 |
|-----------------|-------|-------|-------|
| Pre-`DestructiveTestSuite` r21137 || 5m3s || 4m58s || 6m35s ||
| Post-
|------`DestructiveTestSuite` r21138 || 2m16s || 2m15s || 2m22s ||
| Pre-decorate r22124 | 2m33s | 2m31s | 2m22s |
|---------------------|-------|-------|-------|
| Post-decorate r22125 | 2m48s | 2m49s | 3m3s |
| Head r22997 | 2m55s | 2m42s | 2m51s |
| Head + therve's branch | 2m29s | 2m39s | 2m41s |

Timing command: `time (./bin/trial twisted > /dev/null)`. `real` times reported.

Note that all tests were run on Python 2.5 and I was using my laptop at the time (gmail mostly).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_19"></a>|@jml commented|
|-|-|

Now, to interpret these results.

My reading is that r22125 introduces a change that non-trivially increases the run time of the Twisted test suite, and that therve's branch goes some way to reducing the run time of the Twisted test suite.

I suggest that this ticket cease to be about a performance regression. The focus should be on landing therve's performance improvements. Once they land, I think we can safely close this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_20"></a>|@jml commented|
|-|-|

Python 2.4 numbers. Method same as before.

| Twisted Version | Run 1 | Run 2 | Run 3 |
|-----------------|-------|-------|-------|
| Pre-`DestructiveTestSuite` r21137 || 5m18s || 5m20s || 5m18s ||
| Post-
|------`DestructiveTestSuite` r21138 || 2m21s || 2m19s || 2m26s ||
| Pre-decorate r22124 | 2m33s | 2m37s | 2m22s |
|---------------------|-------|-------|-------|
| Post-decorate r22125 | 3m6s | 3m12s | 3m10s |
| Head r22997 | 3m0s | 2m48s | 2m43s |
| Head + therve's branch | 2m31s | 2m36s | 2m38s |

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml commented|
|-|-|

Proposed conclusion from results: Python 2.4 is not significantly different to Python 2.5 for the purposes of this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

I didn't agree with exarkun about the factor 2 on whole twisted suite. I mainly focused on the force-gc suite, where the regression is. You may not care, but force-gc is a useful feature that helped us prevent bugs in the past, so making it 4 times slower is a problem. The funny thing is that the decorate feature was first added to fix a bug with force-gc!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

Ticket descriptions aren't versioned.  Please be very, very careful when modifying them.  Ideally, don't, just continue to comment.  Sorry, I am not willing to do the work on trac necessary to fix this issue.  We have to live with it unless someone else is going to step up.

In the original ticket description, I said that the runtime of the full test suite _without_ `--force-gc` increased by a factor of 2.  I'm attaching the data file which demonstrates this to be the case.

Here's the summary, however:

```python
>>> sum(a) / len(a)
246.10343373493987
>>> sum(b) / len(b)
503.36140531561472
>>> 503 / 246.
2.0447154471544717
>>>
```

The attached file has three columns of data.  The first is a build number, the second is the 
number of unit tests run in that build, the third is the number of seconds the build took.  Every
data line appears twice for some reason I can't remember.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

Please ignore my last comment.  Can we please just resolve this ticket?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_26"></a>|@jml commented|
|-|-|

OK. I just verified the --force-gc slowdown and therve's fix on my laptop.

Next up, I'll review the code and probably make the fixes myself.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_27"></a>|@jml commented|
|-|-|

(In [23036]) Branching to 'trial-performance-3034-2'

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_28"></a>|@fijal commented|
|-|-|

Instead of storing references to tests, trial might keep weakref to this test. This will allow to access the test as long as it's alive (not sure if this is long enough though), but will not leak references to this test. At least it would not involve horrible cpython-specific hacks. Just random idea.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_29"></a>|@jml removed owner|
|-|-|

OK. I've built something on therve's branch. Here are the differences:

 * Uses a `_clearSuite` function instead of extending the Twisted `TestSuite` interface.
 * Backs out the `DocTestSuite` changes. I'm really sorry about this therve, but I couldn't figure out what it was there for. I'll happily re-introduce them if you can explain them to me.
 * Gives `decorate` an `inPlace` option. On the one hand, it's a YAGNI, on the other hand, I think it's very useful to let people decorate their tests *without* mutating the suite.
 * The refcounting test is protected in case it's running on a system without `sys.getrefcount`.

I didn't add tests to show that `TrialRunner` calls `decorate` with `inPlace=True`. This is partly because it's hard to make a `TrialRunner`, and partly because the _real_ behaviour we care about is the test runtime, not the test reference counting. If the reviewer insists that we test this, I think the best thing to do would be to rephrase the refcounting test in terms of `TrialRunner`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_30"></a>|@therve commented|
|-|-|

test_doctest seems to have an unrelated change, that should be reverted.

The branch fix the problem as expected, I ran it on all the supported slaves.

But I think you failed to address exarkun's main point : "the _tests mangling needs to go through a public API". `_clearSuite` is definitely not a public API. FWIW, that explains my changes to `DocTestSuite`: if you want to add a method on that object, you have to make the changes that I made.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_31"></a>|@jml commented|
|-|-|

OK, that explains the changes to `DocTestSuite`.

In an ideal world, the `_tests` mangling would go through a public API. However, in this world, we need to work with `TestSuite` classes that we don't control. I think that `_clearSuite` is the clearest way of expressing this.

Will fix the unrelated change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

We don't have to control all `TestSuite` classes in order to go through a public API _in the decorate function_.  This is a prime example of an adaptor use-case.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_33"></a>|@jml commented|
|-|-|

We don't yet have an adapter for `TestSuite`. I guess you are suggesting that we should? I'm not convinced one way or the other on that.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_34"></a>|@jml commented|
|-|-|

```
<fijal> http://paste.pocoo.org/show/36038
<fijal> jml: what about this one?
<bigdog> exarkun: good night, just think the muffin shop will get up early to have a nice warm muffin waiting for you 
<fijal> tests the same thing
 but will be portable instead
 FireMoth fijal
<jml> fijal: does it?
<fijal> sorry, you need to delete stuff
 del suite
 del test probably
 also
 if decoratedTest keeps alive suite
 and suite keeps test alive
 getrefcount will not be increased
 and this test is bogus
 you really need to check whether stuff goes away, not exactly number of references being equal
 do I make any sense?
 FireMoth fijal
<radix> +1
* jml pastes into ticket
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_35"></a>|@jml commented|
|-|-|

From http://paste.pocoo.org/show/36038:

```python
import weakref
import gc
test = unittest.TestCase()
suite = unittest.TestSuite([test])
weaktest = weakref.ref(test)
decoratedTest = unittest.decorate(suite, unittest.TestDecorator)
gc.collect()
assert weaktest() is None
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_36"></a>|@radix commented|
|-|-|

That example of testing garbage collection is incorrect.


```
a = something()
weaka = weakref.ref(a)
doSomething(a)
del a
gc.collect()
assert weaka() is None
```

is how this idiom works.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_37"></a>|@therve commented|
|-|-|

Unfortunately, the weakref approach doesn't work here (we don't have a way to cleanly delete the objects).

But the adapter problem is still here...

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_38"></a>|@jml commented|
|-|-|

I am not going to define an interface and add an adapter until there's more than one change we want from `TestSuite`'s API.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_39"></a>|@glyph set owner to @jml|
|-|-|

I think this is fine for now, but I don't like this implementation strategy; as such, I don't want to expose the `inPlace` implementation detail as a public API.  Please remove the flag (as you say, it's YAGNI) and merge.

(I'll have a more useful suggestion for an alternative implementation strategy elsewhere, but there has been too much discussion on this ticket already.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_40"></a>|@therve commented|
|-|-|

(In [23382]) Remove inPlace argument.

Refs [#3034](https://github.com/twisted/twisted/issues/3034)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_41"></a>|@therve commented|
|-|-|

(In [23383]) Branching to 'trial-performance-3034-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_42"></a>|@therve set status to `closed`|
|-|-|

(In [23386]) Merge trial-performance-3034-3

Authors: jml, therve
Reviewer: glyph
Fixes [#3034](https://github.com/twisted/twisted/issues/3034)

Modify twisted.trial.unittest.decorate to modify test suites in-place, to
prevent from a memory problem mainly visible with the force-gc flag.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
